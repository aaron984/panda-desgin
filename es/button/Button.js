import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import classnames from 'classnames';
import React from 'react';
import TapFeedback from 'react-tap-feedback';
import Icon from '../Icon';

var Button =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(Button, _React$PureComponent);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _getPrototypeOf(Button).apply(this, arguments));
  }

  _createClass(Button, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          prefixCls = _this$props.prefixCls,
          type = _this$props.type,
          size = _this$props.size,
          inline = _this$props.inline,
          disabled = _this$props.disabled,
          icon = _this$props.icon,
          loading = _this$props.loading,
          activeStyle = _this$props.activeStyle,
          activeClassName = _this$props.activeClassName,
          onClick = _this$props.onClick,
          restProps = _objectWithoutProperties(_this$props, ["children", "className", "prefixCls", "type", "size", "inline", "disabled", "icon", "loading", "activeStyle", "activeClassName", "onClick"]);

      var iconType = loading ? 'loading' : icon;
      var wrapCls = classnames(prefixCls, className, (_classnames = {}, _defineProperty(_classnames, "".concat(prefixCls, "--primary"), type === 'primary'), _defineProperty(_classnames, "".concat(prefixCls, "--ghost"), type === 'ghost'), _defineProperty(_classnames, "".concat(prefixCls, "--warning"), type === 'warning'), _defineProperty(_classnames, "".concat(prefixCls, "--small"), size === 'small'), _defineProperty(_classnames, "".concat(prefixCls, "--inline"), inline), _defineProperty(_classnames, "".concat(prefixCls, "--disabled"), disabled), _defineProperty(_classnames, "".concat(prefixCls, "--loading"), loading), _defineProperty(_classnames, "".concat(prefixCls, "--icon"), !!iconType), _classnames));
      var iconEl;

      if (typeof iconType === 'string') {
        iconEl = React.createElement(Icon, {
          "aria-hidden": "true",
          type: iconType,
          size: size === 'small' ? 'xxs' : 'md',
          className: "".concat(prefixCls, "-icon")
        });
      } else if (iconType) {
        var rawCls = iconType.props && iconType.props.className;
        var cls = classnames('panda-icon', "".concat(prefixCls, "-icon"), size === 'small' ? 'panda-icon-xxs' : 'panda-icon-md');
        iconEl = React.cloneElement(iconType, {
          className: rawCls ? "".concat(cls, " ").concat(rawCls) : cls
        });
      }

      return React.createElement(TapFeedback, {
        activeClassName: activeClassName || "".concat(prefixCls, "--active"),
        disabled: disabled,
        activeStyle: activeStyle
      }, React.createElement("a", _extends({
        className: wrapCls
      }, restProps, {
        onClick: disabled ? undefined : onClick,
        "aria-disabled": disabled
      }), iconEl, children));
    }
  }]);

  return Button;
}(React.PureComponent);

_defineProperty(Button, "defaultProps", {
  prefixCls: 'panda-button',
  size: 'large',
  inline: false,
  disabled: false,
  loading: false
});

export default Button;