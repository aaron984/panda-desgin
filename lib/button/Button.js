"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _reactTapFeedback = _interopRequireDefault(require("react-tap-feedback"));

var _Icon = _interopRequireDefault(require("../Icon"));

var Button =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2["default"])(Button, _React$PureComponent);

  function Button() {
    (0, _classCallCheck2["default"])(this, Button);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Button).apply(this, arguments));
  }

  (0, _createClass2["default"])(Button, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          prefixCls = _this$props.prefixCls,
          type = _this$props.type,
          size = _this$props.size,
          inline = _this$props.inline,
          disabled = _this$props.disabled,
          icon = _this$props.icon,
          loading = _this$props.loading,
          activeStyle = _this$props.activeStyle,
          activeClassName = _this$props.activeClassName,
          onClick = _this$props.onClick,
          restProps = (0, _objectWithoutProperties2["default"])(_this$props, ["children", "className", "prefixCls", "type", "size", "inline", "disabled", "icon", "loading", "activeStyle", "activeClassName", "onClick"]);
      var iconType = loading ? 'loading' : icon;
      var wrapCls = (0, _classnames2["default"])(prefixCls, className, (_classnames = {}, (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--primary"), type === 'primary'), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--ghost"), type === 'ghost'), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--warning"), type === 'warning'), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--small"), size === 'small'), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--inline"), inline), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--disabled"), disabled), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--loading"), loading), (0, _defineProperty2["default"])(_classnames, "".concat(prefixCls, "--icon"), !!iconType), _classnames));
      var iconEl;

      if (typeof iconType === 'string') {
        iconEl = _react["default"].createElement(_Icon["default"], {
          "aria-hidden": "true",
          type: iconType,
          size: size === 'small' ? 'xxs' : 'md',
          className: "".concat(prefixCls, "-icon")
        });
      } else if (iconType) {
        var rawCls = iconType.props && iconType.props.className;
        var cls = (0, _classnames2["default"])('panda-icon', "".concat(prefixCls, "-icon"), size === 'small' ? 'panda-icon-xxs' : 'panda-icon-md');
        iconEl = _react["default"].cloneElement(iconType, {
          className: rawCls ? "".concat(cls, " ").concat(rawCls) : cls
        });
      }

      return _react["default"].createElement(_reactTapFeedback["default"], {
        activeClassName: activeClassName || "".concat(prefixCls, "--active"),
        disabled: disabled,
        activeStyle: activeStyle
      }, _react["default"].createElement("a", (0, _extends2["default"])({
        className: wrapCls
      }, restProps, {
        onClick: disabled ? undefined : onClick,
        "aria-disabled": disabled
      }), iconEl, children));
    }
  }]);
  return Button;
}(_react["default"].PureComponent);

(0, _defineProperty2["default"])(Button, "defaultProps", {
  prefixCls: 'panda-button',
  size: 'large',
  inline: false,
  disabled: false,
  loading: false
});
var _default = Button;
exports["default"] = _default;