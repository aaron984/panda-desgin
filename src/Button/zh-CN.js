export default () => 'doc页面asdfasdf';