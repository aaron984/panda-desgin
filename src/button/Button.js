import React from 'react';
import PropTypes from 'prop-types';
import TapFeedback from 'react-tap-feedback';
import classnames from 'classnames';
import { isString } from '../utils/checkType';
import Icon from '../icon';

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.element,
  ]).isRequired,
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  activeClassName: PropTypes.string,
  type: PropTypes.string,
  size: PropTypes.string,
  inline: PropTypes.bool,
  loading: PropTypes.bool,
  style: PropTypes.objectOf(PropTypes.string),
  activeStyle: PropTypes.objectOf(PropTypes.string),
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  iconType: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
};

class Button extends React.PureComponent {
  static propTypes = propTypes

  static defaultProps = {
    prefixCls: 'panda-button',
    size: 'large',
    inline: false,
    disabled: false,
    loading: false,
  };

  getIconEl(iconType) {
    if (!iconType) {
      return null;
    }

    const { prefixCls, size } = this.props;
    if (isString(iconType)) {
      return (
        <Icon
          type={iconType}
          size={size === 'small' ? 'xxs' : 'md'}
          className={`${prefixCls}-icon`}
        />
      );
    }

    const rawCls = iconType.props && iconType.props.className;
    const isSmall = size === 'small';
    const cls = classnames({
      'panda-icon': true,
      [`${prefixCls}-icon`]: true,
      'panda-icon-xxs': isSmall,
      'panda-icon-md': !isSmall,
      [rawCls]: rawCls,
    });

    return React.cloneElement(iconType, {
      className: cls,
    });
  }

  render() {
    const { children, className, prefixCls, type, size, inline, disabled,
      icon, loading, activeStyle, activeClassName, onClick, ...restProps
    } = this.props;

    const iconType = loading ? 'loading' : icon;
    const wrapCls = classnames(prefixCls, className, {
      [`${prefixCls}--primary`]: type === 'primary',
      [`${prefixCls}--ghost`]: type === 'ghost',
      [`${prefixCls}--warning`]: type === 'warning',
      [`${prefixCls}--small`]: size === 'small',
      [`${prefixCls}--inline`]: inline,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--loading`]: loading,
      [`${prefixCls}--icon`]: iconType,
    });

    return (
      <TapFeedback
        activeClassName={activeClassName || `${prefixCls}--active`}
        disabled={disabled}
        activeStyle={activeStyle}
      >
        <a
          {...restProps}
          className={wrapCls}
          onClick={disabled ? undefined : onClick}
          aria-disabled={disabled}
        >
          {this.getIconEl(iconType)}
          {children}
        </a>
      </TapFeedback>
    );
  }
}

export default Button;
