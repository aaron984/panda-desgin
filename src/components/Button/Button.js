import classnames from 'classnames';
import React from 'react';
import TapFeedback from 'react-tap-feedback';
import Icon from '../Icon';

class Button extends React.PureComponent {
  static defaultProps = {
    prefixCls: 'panda-button',
    size: 'large',
    inline: false,
    disabled: false,
    loading: false,
    activeStyle: {},
  };

  render() {
    const { children, className, prefixCls, type, size, inline, disabled,
      icon, loading, activeStyle, activeClassName, onClick, ...restProps
    } = this.props;

    const iconType = loading ? 'loading' : icon;
    const wrapCls = classnames(prefixCls, className, {
      [`${prefixCls}-primary`]: type === 'primary',
      [`${prefixCls}-ghost`]: type === 'ghost',
      [`${prefixCls}-warning`]: type === 'warning',
      [`${prefixCls}-small`]: size === 'small',
      [`${prefixCls}-inline`]: inline,
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-icon`]: !!iconType,
    });

    let iconEl;
    if (typeof iconType === 'string') {
      iconEl = (
        <Icon
          aria-hidden="true"
          type={iconType}
          size={size === 'small' ? 'xxs' : 'md'}
          className={`${prefixCls}-icon`}
        />
      );
    } else if (iconType) {
      const rawCls = iconType.props && iconType.props.className;
      const cls = classnames(
        'am-icon',
        `${prefixCls}-icon`,
        size === 'small' ? 'am-icon-xxs' : 'am-icon-md',
      );
      iconEl = React.cloneElement(iconType, {
        className: rawCls ? `${rawCls} ${cls}` : cls,
      });
    }
    return (
      <TapFeedback
        activeClassName={
          activeClassName || (activeStyle ? `${prefixCls}-active` : undefined)}
        disabled={disabled}
        activeStyle={activeStyle}
      >
        <a
          className={wrapCls}
          {...restProps}
          onClick={disabled ? undefined : onClick}
          aria-disabled={disabled}
        >
          {iconEl}
          {children}
        </a>
      </TapFeedback>
    );
  }
}

export default Button;
