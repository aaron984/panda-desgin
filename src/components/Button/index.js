import classnames from 'classnames';
import React from 'react';
import TapFeedback from 'react-tap-feedback';
import Icon from '../icon';

const rxTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;
const isTwoCNChar = rxTwoCNChar.test.bind(rxTwoCNChar);
function isString(str) {
  return typeof str === 'string';
}

function insertSpace(child) {
  if (isString(child.type) && isTwoCNChar(child.props.children)) {
    return React.cloneElement(
      child,
      {},
      child.props.children.split('').join(' '),
    );
  }
  if (isString(child)) {
    if (isTwoCNChar(child)) {
      child = child.split('').join(' ');
    }
    return <span>{child}</span>;
  }
  return child;
}

class Button extends React.Component {
  static defaultProps = {
    prefixCls: 'am-button',
    size: 'large',
    inline: false,
    disabled: false,
    loading: false,
    activeStyle: {},
  };

  render() {
    const {
      children,
      className,
      prefixCls,
      type,
      size,
      inline,
      disabled,
      icon,
      loading,
      activeStyle,
      activeClassName,
      onClick,
      ...restProps
    } = this.props;

    const iconType = loading ? 'loading' : icon;
    const wrapCls = classnames(prefixCls, className, {
      [`${prefixCls}-primary`]: type === 'primary',
      [`${prefixCls}-ghost`]: type === 'ghost',
      [`${prefixCls}-warning`]: type === 'warning',
      [`${prefixCls}-small`]: size === 'small',
      [`${prefixCls}-inline`]: inline,
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-icon`]: !!iconType,
    });

    const kids = React.Children.map(children, insertSpace);

    let iconEl;
    if (typeof iconType === 'string') {
      iconEl = (
        <Icon
          aria-hidden="true"
          type={iconType}
          size={size === 'small' ? 'xxs' : 'md'}
          className={`${prefixCls}-icon`}
        />
      );
    } else if (iconType) {
      const rawCls = iconType.props && iconType.props.className;
      const cls = classnames(
        'am-icon',
        `${prefixCls}-icon`,
        size === 'small' ? 'am-icon-xxs' : 'am-icon-md',
      );
      iconEl = React.cloneElement(iconType, {
        className: rawCls ? `${rawCls} ${cls}` : cls,
      });
    }
    return (
      <TapFeedback
        activeClassName={
          activeClassName || (activeStyle ? `${prefixCls}-active` : undefined)}
        disabled={disabled}
        activeStyle={activeStyle}
      >
        <button
          className={wrapCls}
          {...restProps}
          onClick={disabled ? undefined : onClick}
          aria-disabled={disabled}
        >
          {iconEl}
          {kids}
        </button>
      </TapFeedback>
    );
  }
}

export default Button;
