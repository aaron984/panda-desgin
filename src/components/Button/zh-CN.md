## Button 按钮

### 代码演示

#### 默认
```jsx
<Button>default</Button>
<Button disabled>default disabled</Button>

<Button type="primary">primary</Button>
<Button type="primary" disabled>primary disabled</Button>

<Button type="warning" >warning</Button>
<Button type="warning" disabled >warning disabled</Button>

<Button icon="check-circle-o">with icon</Button>
```

#### small and line
```jsx
<Button loading inline size="small">loading button</Button>
<Button icon="check-circle-o" inline size="small" >with icon and inline</Button>

<Button type="ghost" inline size="small">inline ghost</Button>
<Button type="primary" inline size="small" disabled>primary  disabled</Button>
```

## API

属性 | 说明 | 类型 | 默认值
----|-----|------|------
| type    | 按钮类型，可选值为`primary`/`ghost`/`warning`或者不设  |   string   |   -  |
| size    | 按钮大小，可选值为`large`、`small` | string | `large`|
| activeStyle  | 点击反馈的自定义样式 (设为 false 时表示禁止点击反馈) | {}/false | {} |
| activeClassName  | 点击反馈的自定义类名 | string |  |
| disabled   | 设置禁用  | boolean |    false  |
| onClick    | 点击按钮的点击回调函数 | (e: Object): void |   无  |
| style    | 自定义样式 |   Object  | 无 |
| inline     | 是否设置为行内按钮  | boolean |   false  |
| loading	   | 设置按钮载入状态	  | boolean	 | false |
| icon  | 可以是组件里内置的某个 icon 的 type 值，也可以是任意合法的 ReactElement (注意: `loading`设置后此项设置失效) | `string`, `React.Element` | -  |
| prefixCls |  class前缀 | string | `panda-button` |
| className |  样式类名 | string | 无 |
