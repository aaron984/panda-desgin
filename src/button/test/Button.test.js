import React from 'react';
import { render, mount } from 'enzyme';
import Button from '../index';

describe('Button', () => {
  it('renders correctly', () => {
    const wrapper = render(<Button>Follow</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('configre size equal to should correctly', () => {
    const wrapper = mount(<Button icon="check-circle-o" size="small">Follow</Button>);
    expect(wrapper.find('.panda-button--small')).toHaveLength(1);
  });

  it('configre icon should correctly', () => {
    const wrapper = mount(<Button icon="check-circle-o">Follow</Button>);
    expect(wrapper.find('.pan-icon-check-circle-o')).toHaveLength(1);
  });

  it('configre icon should correctly', () => {
    const wrapper = mount(<Button icon="check-circle-o">Follow</Button>);
    expect(wrapper.find('.pan-icon-check-circle-o')).toHaveLength(1);
  });

  it('icon is node should correctly', () => {
    const wrapper = mount(<Button icon={<span>!</span>}>Follow</Button>);
    expect(wrapper.find('.panda-icon')).toHaveLength(1);
    expect(wrapper.find('.panda-icon-md')).toHaveLength(1);
  });

  it('icon is node and sizw equal to small should correctly', () => {
    const wrapper = mount(<Button icon={<span className="self-icon-class">!</span>} size="small">Follow</Button>);
    expect(wrapper.find('.panda-icon')).toHaveLength(1);
    expect(wrapper.find('.panda-icon-xxs')).toHaveLength(1);
    expect(wrapper.find('.self-icon-class')).toHaveLength(1);
  });

  it('loading should correctly', () => {
    const wrapper = mount(<Button loading>Follow</Button>);
    expect(wrapper.find('.pan-icon-loading')).toHaveLength(1);
    expect(wrapper.find('.panda-button--loading')).toHaveLength(1);
  });

  const click = jest.fn();

  it('disabled should correctly', () => {
    const wrapper = mount(<Button disabled onClick={click}>Follow</Button>);
    expect(wrapper.find('.panda-button--disabled')).toHaveLength(1);
    expect(click).not.toHaveBeenCalled();
  });

  it('click event should correctly', () => {
    const wrapper = mount(<Button onClick={click}>Follow</Button>);
    wrapper.find('a').simulate('click');
    expect(click).toHaveBeenCalled();
  });
});
