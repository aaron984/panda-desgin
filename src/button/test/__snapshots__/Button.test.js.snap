// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "aria-disabled": "false",
      "class": "panda-button",
    },
    "children": Array [
      Object {
        "data": "Follow",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "a",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-disabled": undefined,
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-disabled": undefined,
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
